aws-lc-rs/src/rsa/encryption.rs

```diff
@@ -0,0 +1,510 @@
+    pub fn decrypt<'plaintext>(
+        &self,
+        algorithm: &'static OaepAlgorithm,
+        ciphertext: &[u8],
+        plaintext: &'plaintext mut [u8],
+        label: Option<&[u8]>,
+    ) -> Result<&'plaintext mut [u8], Unspecified> {
+        let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.private_key.0, null_mut()) })?;
+
+        if 1 != unsafe { EVP_PKEY_decrypt_init(*pkey_ctx) } {
+            return Err(Unspecified);
+        }
+
+        configure_oaep_crypto_operation(
+            &pkey_ctx,
+            algorithm.oaep_hash_fn(),
+            algorithm.mgf1_hash_fn(),
+            label,
+        )?;
+
+        let mut out_len = plaintext.len();
+
+        if 1 != indicator_check!(unsafe {
+            EVP_PKEY_decrypt(
+                *pkey_ctx,
+                plaintext.as_mut_ptr(),
+                &mut out_len,
+                ciphertext.as_ptr(),
+                ciphertext.len(),
+            )
+        }) {
+            return Err(Unspecified);
+        };
+
+        Ok(&mut plaintext[..out_len])
+    }
+
+    /// Returns the RSA key size in bytes.
+    #[must_use]
+    pub fn key_size_bytes(&self) -> usize {
+        self.private_key.key_size_bytes()
+    }
+
+    /// Returns the RSA key size in bits.
+    #[must_use]
+    pub fn key_size_bits(&self) -> usize {
+        self.private_key.key_size_bits()
+    }
+
+    /// Returns the minimum plaintext buffer size required for `Self::decrypt`.
+    #[must_use]
+    pub fn min_output_size(&self) -> usize {
+        self.key_size_bytes()
+    }
+}
+
+impl Debug for OaepPrivateDecryptingKey {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("OaepPrivateDecryptingKey")
+            .finish_non_exhaustive()
+    }
+}
+
+fn configure_oaep_crypto_operation(
+    evp_pkey_ctx: &LcPtr<EVP_PKEY_CTX>,
+    oaep_hash_fn: OaepHashFn,
+    mgf1_hash_fn: Mgf1HashFn,
+    label: Option<&[u8]>,
+) -> Result<(), Unspecified> {
+    if 1 != unsafe { EVP_PKEY_CTX_set_rsa_padding(**evp_pkey_ctx, RSA_PKCS1_OAEP_PADDING) } {
+        return Err(Unspecified);
+    };
+
+    if 1 != unsafe { EVP_PKEY_CTX_set_rsa_oaep_md(**evp_pkey_ctx, oaep_hash_fn()) } {
+        return Err(Unspecified);
+    };
+
+    if 1 != unsafe { EVP_PKEY_CTX_set_rsa_mgf1_md(**evp_pkey_ctx, mgf1_hash_fn()) } {
+        return Err(Unspecified);
+    };
+
+    let label = if let Some(label) = label {
+        label
+    } else {
+        &[0u8; 0]
+    };
+
+    if label.is_empty() {
+        // Safety: Don't pass zero-length slice pointers to C code :)
+        if 1 != unsafe { EVP_PKEY_CTX_set0_rsa_oaep_label(**evp_pkey_ctx, null_mut(), 0) } {
+            return Err(Unspecified);
+        }
+        return Ok(());
+    }
+
+    // AWS-LC takes ownership of the label memory, and will call OPENSSL_free, so we are forced to copy it for now.
+    let label_ptr =
+        DetachableLcPtr::<u8>::new(unsafe { OPENSSL_malloc(size_of_val(label)) }.cast())?;
```

Comment:
NP: is it possible o avoid this duplication and tell rust to release the original label that was passed in?

Reply:
No there is not, and on the Rust side since we take a reference to the label content (`&[u8]`) that could actually just be a fixed slice that is `'static` lifetime so there is nothing that needs to be freed. Problem is there is no way to tell AWS-LC that we own the allocation and stop it from trying to free the memory pointer too.